Add eva driver script

kevindougherty-noaa opened the pull request:
This PR introduces the first implementation of the infrastructure required to run a driver script for observation monitoring. The current version includes the following functionality:

- Copies NetCDF files from a specified experiment_dir to a runtime directory.

- Dynamically generates an EVA configuration file using a Jinja2 template.

- Executes the EVA process with the generated configuration.

The copy-to-output-directory logic is currently commented out until we make a decision on where we want to point these figures. We will eventually support multiple templates to for all the figures we want to generate.

Some paths and parameters are still hardcoded and would benefit from a more flexible configuration or templating approach. Feedback and suggestions on how to improve this are welcome.

kevindougherty-noaa committed:
Add driver scripts. First pass will need reviews

CoryMartin-NOAA commented on driver/config.yaml:
```diff
@@ -0,0 +1,9 @@
+experiment_dir: '/scratch1/NCEPDEV/da/Kevin.Dougherty/test/'
```
not needed now, but if we ever 'operationalize' this, we will need to use the standard naming conventions for directories.

CoryMartin-NOAA replied:
We should consider adding for each ob type a template relative path to the input files, and replace dates, rather than just globbing

CoryMartin-NOAA replied:
```
template_path = './gdas.%Y%m%d/%H/products/obstats/gdas.t%Hz.viirs_n20_iodastats.nc`
os.path.join(experiment_dir, filledin_template_dir)
```

kevindougherty-noaa replied:
So we would need to point to a specific cycle to do this, correct?

CoryMartin-NOAA replied:
If I understand correctly, yes, we need to define which cycle(s) we want to read in

CoryMartin-NOAA commented on driver/config.yaml:
```diff
@@ -0,0 +1,9 @@
+end: '2023-11-30T18:00:00Z'
+interval_hours: 6
+outdir: '/scratch1/NCEPDEV/da/Kevin.Dougherty/outdir'
+ob type: 'viirs_n20'
+plot type: 'timeseries'
+template_path: '/scratch1/NCEPDEV/da/Kevin.Dougherty/obs-monitor/templates/ioda_stats_viirs_n20.yaml.j2'
```
Already at this stage, we should make this a list of dictionaries so that you can process multiple ob types at once.

CoryMartin-NOAA commented on driver/driver.py:
```diff
@@ -0,0 +1,130 @@
+        raise FileNotFoundError(f"Template path not found: {template_path}")
+
+    # Create runtime directory
+    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
```
we should get PDY and cyc from the environment to get the cycle time and not the exact current UTC time, will make it easier to find what we are looking for

CoryMartin-NOAA commented on driver/driver.py:
```diff
@@ -0,0 +1,130 @@
+
+    # Create runtime directory
+    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
+    runtime_dir = runtime_root / f"runtime_{timestamp}"
```
consider os.path.join to make it clearer to the reader

CoryMartin-NOAA commented on driver/driver.py:
```diff
@@ -0,0 +1,130 @@
+    runtime_dir.mkdir(parents=True, exist_ok=False)
+    logging.info(f"Created runtime directory: {runtime_dir}")
+
+    # Copy experiment files
+    ob_type = monitor_dict["ob type"]
+    nc_files = list(experiment_dir.glob(f"*{ob_type}_*.nc"))
+    if not nc_files:
+        logging.warning(f"No NetCDF files found matching *{ob_type}_*.nc in {experiment_dir}")
+    for file in nc_files:
+        shutil.copy2(file, runtime_dir / file.name)
+        logging.info(f"Copied: {file.name}")
```
We should only look for and copy files in the start and end time period that is defined

CoryMartin-NOAA commented on driver/driver.py:
```diff
@@ -0,0 +1,130 @@
+    runtime_dir = runtime_root / f"runtime_{timestamp}"
+    runtime_dir.mkdir(parents=True, exist_ok=False)
+    logging.info(f"Created runtime directory: {runtime_dir}")
+
```
Basically from here down should be a parallelized loop over ob types

CoryMartin-NOAA commented on driver/driver.py:
```diff
@@ -0,0 +1,130 @@
+        logging.error(f"EVA failed with exit code {e.returncode}")
+        raise
+
+    # Copy results to output directory
```
as you said but just noting here, this should copy/move files to a directory and naming structure that aligns with the web pages @EdwardSafford-NOAA is working on

EdwardSafford-NOAA replied:
Here's an initial cut at the desired directory structure.  This isn't a directive by any means; I'm open to any changes that make sense, now or later.  

The top level directory, 'public', corresponds to the structure of the website in React.  If you dump the files into a structure like this on the dev machine then the local public directory can be rsync-ed to the public directory on the website. 

This example assumes that we're just making summary and time series plots.  It can be expanded if we opt to do more/different plots in the standard set.

For radiance plots:

public/
└── plots/
    └── [satellite]/
        └── [instrument]/
            └── [channel_[channelNumber].png
            └── summary.png
             
 Ozone plots can be the same except skip the channel.
 
 Conventional plots 
public/
└── plots/
    └── [Conv Type]*
        └── [type-subtype]**
            └── [type-subtype.png
            
* Conv Type = gps, ps, q, t, uv
** ConType type-subtype = 180-0, 242-55, etc.

CoryMartin-NOAA reviewed (commented):
First round of comments, we can discuss in 40 mins, @kevindougherty-noaa

kevindougherty-noaa committed:
checkpoint for adding multiprocessing. Need to update config files

kevindougherty-noaa committed:
add different example of config file

kevindougherty-noaa committed:
update config file to use input cycle as end date and calculate start date

kevindougherty-noaa committed:
Clean up templates with more variables

kevindougherty-noaa committed:
adding first pass at higher level scripts to run driver

kevindougherty-noaa committed:
save progress

kevindougherty-noaa committed:
working generate_workflow_xml.py

kevindougherty-noaa committed:
clean up files

kevindougherty-noaa committed:
save point

kevindougherty-noaa committed:
add shell script

kevindougherty-noaa committed:
edit tempalte

kevindougherty-noaa committed:
save point

kevindougherty-noaa committed:
update main config file

kevindougherty-noaa committed:
add updated files

kevindougherty-noaa committed:
update to proper order in outdir

kevindougherty-noaa committed:
add module file to ursa, similar to GDASApp intel build

kevindougherty-noaa committed:
update scripts and files for successful ursa run

kevindougherty-noaa committed:
update detect_machine.sh for ursa and clean up run_driver.sh

kevindougherty-noaa committed:
update driver for option to remove or keep runtimedir

kevindougherty-noaa committed:
cleanup

kevindougherty-noaa commented:
@CoryMartin-NOAA @EdwardSafford-NOAA @ADCollard I think this is finally at a good spot for official reviews. Our friend put this review together to highlight what this PR is trying to accomplish.

### **Overview**
This PR introduces a new observation monitoring workflow feature. It provides an automated system for monitoring satellite-based observation data, specifically focusing on the ioda-stat files. The workflow is designed to be modular, configurable, and compatible with HPC environments via Rocoto workflow management.

### **Key Components**
**Driver and Configuration**

`driver/driver.py`: Core Python driver script that orchestrates the workflow.

- Creates timestamped runtime directories.
- Copies relevant NetCDF observation files into the runtime environment.
- Generates configuration files from Jinja2 templates.
- Executes the EVA executable to process data.
- Optionally copies output plots to a public directory and cleans up temporary files based on environment variables.

`driver/config.yaml`: Defines monitoring jobs for specific satellites and sensors with paths to their respective Jinja templates.

**Rocoto Workflow Automation**

`parm/generate_workflow_xml.py`: Script to generate the Rocoto XML workflow file from a Jinja2 template, parameterized for flexible runs.

`parm/monitor_rocoto_template.xml.j2`: Jinja2 template for the Rocoto XML workflow that defines task parameters, job scheduling, and environment variables.

`jobs/rocoto/run_driver.sh`: Shell script that loads machine-specific modules and runs the main Python driver.

**Environment and Module Management**

`modulefiles/obs-monitor/ursa.lua`: Lua modulefile that loads the necessary software environment and Python virtual environment, ensuring consistency across different HPC platforms.

`ush/detect_machine.sh`: Added ursa capabilities from GDASApp

**Templates**

`templates/ioda_stats_viirs_n20.yaml.j2` and `templates/ioda_stats_viirs_npp.yaml.j2`: Jinja2 configuration templates specifying dataset definitions and plotting instructions for VIIRS N20 and NPP satellites, respectively. These templates configure how data is processed and visualized using Emcpy plotting backend.

### **Design Highlights**
Uses environment variables extensively to allow flexible runtime configuration (e.g., experiment directories, intervals, number of cycles, data retention policies).

Modular design supports multiple satellites/sensors by specifying them in a single config YAML (driver/config.yaml).

Robust handling of machine-specific environments via the module system and detection script to ensure portability across HPC systems like Hera, Ursa, WCOSS2, etc.

Automated workflow generation for Rocoto simplifies scheduling and integration into existing HPC batch environments.

Includes careful file management with options to keep or clean up runtime data and to publish generated plots to a public directory.

CoryMartin-NOAA commented on driver/driver.py:
```diff
@@ -0,0 +1,181 @@
+from wxflow import Logger, Jinja
+from wxflow.configuration import cast_as_dtype
+
+def generate_eva_config(template_path: str, output_path: str, context: dict):
```
Do we really need this function? isn't this a duplication of something in wxflow?

kevindougherty-noaa replied:
It is not a replication of something in wxflow, it uses wxflow's capabilities. If you dont think it needs to be a function I will just add it back into the main driver function.

CoryMartin-NOAA replied:
My mistake, it's fine, I guess the question is do you envision using it more places or just in the one spot? If just the one spot, I'd just replace the function call with the two lines

CoryMartin-NOAA commented on driver/driver.py:
```diff
@@ -0,0 +1,181 @@
+    """
+    monitor_dict, timestamp = args
+
+    satellite = monitor_dict["satellite"]
+    sensor = monitor_dict["sensor"]
+    ob_type = f"{sensor}_{satellite}"
```
This is not going to be generic enough as an FYI, like for conventional/in situ obs

CoryMartin-NOAA commented on driver/driver.py:
```diff
@@ -0,0 +1,181 @@
+    runtime_root = Path(os.getenv("RUNTIME_DIR"))
+    template_path = monitor_dict["template_path"]
+
+    runtime_dir = runtime_root / f"runtime_{ob_type}_{timestamp}"
```
os.path.join

kevindougherty-noaa replied:
I would argue that the current implementation using a Path() object with the forward slash is more modern and preferred and in my opinion more readable by a user than `os.path.join()`. Performance wise is pretty much negligible difference as well.

CoryMartin-NOAA replied:
Ok you've got a convincing enough argument

CoryMartin-NOAA commented on jobs/rocoto/run_driver.sh:
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env bash
```
for ops, we need this script to call a J-job in jobs, and then that script needs to call an ex-script in scripts/ which can be a python script

EdwardSafford-NOAA replied:
There is a j-job and an ex-script in obs-monitor.  They can be modified as needed. They are in /jobs and /scripts directories respectively.  Also there's a driver/runObsMon.sh, a test script which submits the j-job.

CoryMartin-NOAA commented on modulefiles/obs-monitor/ursa.lua:
```diff
@@ -0,0 +1,54 @@
+  prepend_path("PATH", "/scratch3/NCEPDEV/da/Edward.Safford/noscrub/python/envs/obs-mon/bin")
+  prepend_path("PYTHONPATH", "/scratch3/NCEPDEV/da/Edward.Safford/noscrub/python/envs/obs-mon")
+
+else
```
One of my AI queries a few days ago mentioned that instead of `activate` and `deactivate` that there was a better way to use a python virtual environment when loading a module but I forget the details. May be worth asking about it to a "friend"

CoryMartin-NOAA commented on parm/generate_workflow_xml.py:
parm should not have scripts in it, this should go in somewhere like dev/workflow/

CoryMartin-NOAA commented on templates/ioda_stats_viirs_n20.yaml.j2:
```diff
@@ -0,0 +1,45 @@
+            slices: '[:,0]'
+          markersize: 5
+          color: 'black'
+          label: 'aerosolOpticalDepth'
```
No EOL

CoryMartin-NOAA commented on templates/ioda_stats_viirs_npp.yaml.j2:
```diff
@@ -0,0 +1,45 @@
+            slices: '[:,0]'
+          markersize: 5
+          color: 'black'
+          label: 'aerosolOpticalDepth'
```
No EOL

CoryMartin-NOAA commented on ush/detect_machine.sh:
```diff
@@ -67,26 +74,34 @@ if [[ -d /lfs/h3 ]]; then
 else
   echo WARNING: UNKNOWN PLATFORM 1>&2
-fi
+fi
```
Missing EOL

kevindougherty-noaa committed:
update minor fixes

kevindougherty-noaa committed:
update to not activate venv, but points to correct env to run

kevindougherty-noaa commented on parm/generate_workflow_xml.py:
```diff
@@ -0,0 +1,61 @@
+        SDATE=args.start_date,
+        EDATE=args.end_date,
+        ACCOUNT="da-cpu",
+        QUEUE="batch",
```
Curious about these two reviews. Should I have like a config file similar to g-w and point variables like queue, account, etc.?

CoryMartin-NOAA replied:
that, or you can just have command line args with default values if not specified?

kevindougherty-noaa committed:
move generate_workflow_xml.py

kevindougherty-noaa commented on driver/driver.py:
```diff
@@ -0,0 +1,181 @@
+    template_path = monitor_dict["template_path"]
+
+    runtime_dir = runtime_root / f"runtime_{ob_type}_{timestamp}"
+    runtime_dir.mkdir(parents=True, exist_ok=False)
```
I am addressing this one currently, but I always had a little concern about this because of potential overlap. I don't want jobs that could be similar copying data over into a runtime_dir that potentially could get deleted. Any ideas to best deter this? Chat suggested `uuid`, I was just thinking of adding a job name in the config file so it is specific. Thoughts?

CoryMartin-NOAA replied:
so, the way the global-workflow handles this is that runtime directories are usually something like:
/path/to/stmp/jobname.12345 where 12345 is the slurm/pbs job ID so it is somewhat unique

kevindougherty-noaa replied:
I looked into this a bit. Sbatch job id's can actually get hairy because we are running multiprocessing in the sbatch submission so each runtime_dir would not be unique. UUID is very simple and always a unique code.

```
import uuid

unique_id = uuid.uuid4()  # Generates a random UUID
print(unique_id)  # Example: '4e0c3b8a-fdea-4f48-a403-349b74ad2e41'
```

So every job submission would have a new id and we could just grab the first or last few digits and not use the entire long id. Thoughts?

CoryMartin-NOAA replied:
uuid works for me then

kevindougherty-noaa committed:
revert output path name becasue eva fails

kevindougherty-noaa committed:
refactor driver.py. Cleaner and more modular

kevindougherty-noaa committed:
fix docstring

kevindougherty-noaa commented:
@CoryMartin-NOAA @EdwardSafford-NOAA @ADCollard I updated driver.py with the help of our friend. I mentioned it to Cory, but I prompted it to review my initial code as a senior developer and actually gave some phenomonal feedback. The new code is much cleaner, more modular, and easier to follow.

- All environment variables are read in the through the MonitoringConfig class which is much cleaner
- Finds matching .nc files based on the start and end times and uses regular expression to handle datetimes in filenames more robustly
- copies .nc files to runtime directory that has a UUID added to directory name so there can be no overlap in jobs
- combines and cleans up eva config path related things and runs eva

I will note this does not handle conventional data yet. Do you think we should get this through and get back to that when we generate the ioda_stat p files we discussed last week or do it now?

CoryMartin-NOAA commented:
I would suggest doing it now as it may lead to modifying the class to handle more types of data. Unless it will be a lot of effort

kevindougherty-noaa commented:
@CoryMartin-NOAA my only pushback is that I can't test it currently. I can open an issue outlining the problem so we don't forget it, but that way it's just a simple implementation down the line. However, we could add something to the config where the jobs are:

```
jobs:
  - monitor_type: radiance 
    satellite: 'n20'
    sensor: 'viirs'
    template_path: '${HOMEobsmon}/templates/ioda_stats_viirs_n20.yaml.j2'

  - monitor_type: radiance  
    satellite: 'npp'
    sensor: 'viirs'
    template_path: '${HOMEobsmon}/templates/ioda_stats_viirs_npp.yaml.j2'
    
 - monitor_type: conventional
    variable: p
    template_path: '${HOMEobsmon}/templates/ioda_stats_conv_p.yaml.j2'
```

And then in the new class have some type of logic:

```
class MonitoringConfig:
    """
    Encapsulates configuration for a single observation monitoring job.
    Handles environment variables, time window calculation, and paths.
    """
    def __init__(self, monitor_dict: dict, timestamp: str):
        if monitor_dict['monitor_type'] == 'radiance':
            self.satellite = monitor_dict["satellite"]
            self.sensor = monitor_dict["sensor"]
            self.ob_type = f"{self.sensor}_{self.satellite}"
        elif monitor_dict['monitor_type'] == 'conventional':
            self.variable = monitor_dict["variable"]
            self.ob_type = f"conv_{self.variable}"
```

And then adjust anything that needs those variables with conditionals i.e the `get_jinja_contex()` method that populates the template jinja files. The most important thing is the `ob_type` matches whats on the ioda-stats .nc file.

CoryMartin-NOAA commented:
we have totalSnowDepth observations that can be used to test it, as those are "conventional", meaning they have no channel dimension

kevindougherty-noaa committed:
reconfigure scripts to add conventional obs

kevindougherty-noaa commented:
Refactored the driver to handle start and end dates better as well as handle conventional obs. Updated the xml generator as well. The changes were not intrusive in the actual driver, more so just updating config to point to the correct monitor type. Tested both and they work as expected.

kevindougherty-noaa committed:
generate_workflow_xml.py

kevindougherty-noaa committed:
add snowcvr template

kevindougherty-noaa committed:
Merge branch 'feature/eva_driver' of https://github.com/NOAA-EMC/obs-monitor into feature/eva_driver

kevindougherty-noaa committed:
update new design to only look for fils in gdas.PDY dir in DATAROOT

kevindougherty-noaa committed:
add new feature that only runs jobs for components listed rather than all jobs

kevindougherty-noaa committed:
update component argparse to work appropriately with new script changes

kevindougherty-noaa committed:
clean up and run successful tests

kevindougherty-noaa committed:
fix typo

kevindougherty-noaa commented:
@EdwardSafford-NOAA @CoryMartin-NOAA I have updated and successfully run tests for both datasets from a clean clone of this branch. The test data exists in `/scratch3/NCEPDEV/da/Kevin.Dougherty/test`. I have two days of snow data (`gdas.20250131` and `gdas.20250201`) and one day of aod data (`gdas.20231130`).

The main changes have been adding a variable `component` which is now required when building the xml file. This tells you which type of jobs you want to run i.e. atmos, snow, ocean, chem, ice. This grabs information from the `config.yaml` in the driver directory so we don't have failing jobs if we are running just snow or just aod because of different dates. If we had data on the same day, we can add these as a single string i.e. "snow,chem" and it would be converted into a list and create the jobs.

The driver also now requires that all data be in the format that we discussed over email: `$DATAROOT/$RUN.$PDY/$CYC/products/$COMPONENT/anlmon/*.nc`

Here is the updated documentation. Please give it a try whenever you get a chance!

Documentation: https://docs.google.com/document/d/1dAivmvtHHh0yQ5Y5D85HpIRNRopP-DXIsVbB0uHbFtE/edit?tab=t.0

CoryMartin-NOAA commented:
@kevindougherty-noaa remind me where we are on this. Is this ready for a thorough review before merging? I think it's been tested now by @EdwardSafford-NOAA right? After finding/solving documentation/instructions issues?

kevindougherty-noaa commented:
@CoryMartin-NOAA I have received some feedback from @EdwardSafford-NOAA that I need to support handling for missing files, but when all files were present it worked. Will be on that this week.

kevindougherty-noaa committed:
update for better logging, add empty stub nc file

kevindougherty-noaa committed:
clean up

kevindougherty-noaa committed:
update templates and requirements.txt

kevindougherty-noaa committed:
copilot reviews

CoryMartin-NOAA reviewed (approved):
let's get this merged and go from there after your leave
